### PR TITLE
Allow preserving headers on context stash

### PR DESCRIPTION
**Pull request by n1v0lg**
This PR adds functionality to stash a thread context but preserve specific request headers. The [need](https://github.com/elastic/elasticsearch/pull/94665#discussion_r1146126754) for this method most recently came up in the context of work on the remote cluster security model.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/util/concurrent/ThreadContext.java [hunk 1]
```diff
@@ -102,17 +105,38 @@ public ThreadContext(Settings settings) {
...
+     * if these exist in the context before stashing.
+     */
+    public StoredContext stashContextPreservingRequestHeaders(final Set<String> requestHeaders) {
+        final Set<String> allHeadersToCopy = new HashSet<>(requestHeaders);
+        allHeadersToCopy.addAll(HEADERS_TO_COPY);
+        return stashContext(Set.copyOf(allHeadersToCopy));
+    }
+
+    public StoredContext stashContextPreservingRequestHeaders(final String... requestHeaders) {
+        return stashContextPreservingRequestHeaders(Arrays.stream(requestHeaders).collect(Collectors.toUnmodifiableSet()));
```
Comment (ywangd) [inline server/src/main/java/org/elasticsearch/common/util/concurrent/ThreadContext.java:122]: Nit:
```suggestion
        return stashContextPreservingRequestHeaders(Set.of(requestHeaders));
```

**Review by ywangd — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/common/util/concurrent/ThreadContext.java [hunk 2]
```diff
@@ -102,17 +105,38 @@ public ThreadContext(Settings settings) {
...
          * X-Opaque-ID should be preserved in a threadContext in order to propagate this across threads.
          * This is needed so the DeprecationLogger in another thread can see the value of X-Opaque-ID provided by a user.
-         * The same is applied to Task.TRACE_ID.
-         * Otherwise when context is stashed, it should be empty.
+         * The same is applied to Task.TRACE_ID and other values specified in `HEADERS_TO_COPY`.
          */
         boolean hasHeadersToCopy = false;
         if (context.requestHeaders.isEmpty() == false) {
-            for (String header : HEADERS_TO_COPY) {
+            for (String header : requestHeadersToCopy) {
```
[the remark this answers is not in the document]
Comment (n1v0lg) [reply server/src/main/java/org/elasticsearch/common/util/concurrent/ThreadContext.java:139]: Good point. I'll make the necessary changes.

**Review by pgomulka — APPROVED** [review]
LGTM

**Comment by n1v0lg** [comment]
@elasticmachine update branch

**Comment by n1v0lg** [comment]
@elasticmachine update branch

**Pull request merged**